src/MonitorProfiler/Utilities/NameCache.cpp

```diff
@@ -0,0 +1,223 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "NameCache.h"
+#include "macros.h"
+#include "corhlpr.h"
+
+const tstring NameCache::CompositeClassName = _T("_CompositeClass_");
+const tstring NameCache::ArrayClassName = _T("_ArrayClass_");
+const tstring NameCache::UnknownName = _T("_Unknown_");
+const tstring NameCache::ModuleSeparator = _T("!");
+const tstring NameCache::FunctionSeperator = _T(".");
+const tstring NameCache::NestedSeparator = _T("+");
+const tstring NameCache::GenericBegin = _T("<");
+const tstring NameCache::GenericSeparator = _T(",");
+const tstring NameCache::GenericEnd = _T(">");
+
+bool NameCache::GetFunctionData(FunctionID id, std::shared_ptr<FunctionData>& data)
+{
+    return GetData(_functionNames, id, data);
+}
+bool NameCache::GetClassData(ClassID id, std::shared_ptr<ClassData>& data)
+{
+    return GetData(_classNames, id, data);
+}
+bool NameCache::GetModuleData(ModuleID id, std::shared_ptr<ModuleData>& data)
+{
+    return GetData(_moduleNames, id, data);
+}
+bool NameCache::GetTokenData(ModuleID modId, mdTypeDef token, std::shared_ptr<TokenData>& data)
+{
+    auto const& it = _names.find(std::make_pair(modId, token));
+
+    if (it != _names.end())
+    {
+        data = it->second;
+        return true;
+    }
+    return false;
+}
+
+void NameCache::AddModuleData(ModuleID moduleId, tstring&& name)
+{
+    _moduleNames.emplace(moduleId, std::make_shared<ModuleData>(std::move(name)));
+}
+
+HRESULT NameCache::GetFullyQualifiedName(FunctionID id, tstring& name)
+{
+    HRESULT hr;
+
+    std::shared_ptr<FunctionData> functionData;
+    if (GetFunctionData(id, functionData))
```

Comment:
Presumably, if this evaluates to false, then we skip all of the logic in this method and don't construct `name`. Do we still want to return `S_OK` in this case, or is this considered a failure?